### ISSUE TITLE
Feature Request: consider storing image references in profile as a relative path

### ISSUE THREAD

**Issue by matt-cahill** [DCS-Skunkworks/DCSFlightpanels#56]
This would allow people to share their streamdeck profiles without the recipient having to re-link all of the image assets. The path should be relative to the profile file.

Another option would be to have a collect and export function where all the assets were collected together and a portable version of the profile was generated and delivered to the user as a zip or similar bundle.

**Comment by matt-cahill** [comment]
@jdahlblom I believe this is relevant to this request.

I was getting a weird crash after upgrading from version 4.015 to 4.0.18 when switching layers using a navigation button. When switching layers using the UI the following was written to the log.

```
22.04.2020 07:46:12  version : 1.0.214.2950
20135444 Custom message = []
System.ArgumentException
Parameter is not valid.
   at System.Drawing.Bitmap..ctor(String filename)
   at NonVisuals.StreamDeck.FaceTypeImage.Show()
   at NonVisuals.StreamDeck.FaceTypeBase.set_IsVisible(Boolean value)
   at NonVisuals.StreamDeck.StreamDeckLayerHandler.SetActiveLayer(String layerName)
   at NonVisuals.StreamDeck.StreamDeckLayerHandler.set_ActiveLayer(String value)
   at DCSFlightpanels.PanelUserControls.StreamDeckUserControl.ComboBoxLayers_OnDropDownClosed(Object sender, EventArgs e)
```

After a bit of playing around I realised it was because the layer I was switching to had a Go Home button configured which referenced a png in the 4.0.15 folder that I had just deleted. The following change fixed my profile.

```diff
        "$values": [
          {
            "$type": "NonVisuals.StreamDeck.StreamDeckButton, NonVisuals",
            "StreamDeckButtonName": 1,
            "Face": {
              "$type": "NonVisuals.StreamDeck.FaceTypeImage, NonVisuals",
              "FaceType": 2,
-              "ImageFile": "D:\\DCS Tools\\DCSFlightpanels_x64_4.0.15_alpha\\\\StreamDeckGallery\\Symbols\\home_white.png",
+              "ImageFile": "D:\\DCS Tools\\DCSFlightpanels_x64_4.0.18_alpha\\\\StreamDeckGallery\\Symbols\\home_white.png",
              "StreamDeckButtonName": 1,
              "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}",
              "OffsetX": 0,
              "OffsetY": 0
            },
            "ActionForPress": {
              "$type": "NonVisuals.StreamDeck.ActionTypeLayer, NonVisuals",

**Comment by jdahlblom** [comment]
True, I will switch to relative path for next relase. Coming weekend.

**Comment by matt-cahill** [comment]
I feel like there will need to be some kind of re-linking process or default png to fall back to if there is not an image file at the location in the profile too.

**Comment by jdahlblom** [comment]
Will do, a default "file not found" image of some sort.

**Comment by jdahlblom** [comment]
Check latest alpha.

**Comment by jdahlblom** [comment]
Fixed. Can you close if OK?

**Comment by matt-cahill** [comment]
For me it still saves a fully qualified path in 4.0.22. Are there any caveats e.g. the image needs to be in the same path tree as the file?

```json
{
  "$type": "System.Collections.Generic.List`1[[NonVisuals.StreamDeck.StreamDeckLayer, NonVisuals]], mscorlib",
  "$values": [
    {
      "$type": "NonVisuals.StreamDeck.StreamDeckLayer, NonVisuals",
      "Name": "Home",
      "StreamDeckButtons": {
        "$type": "System.Collections.Generic.List`1[[NonVisuals.StreamDeck.StreamDeckButton, NonVisuals]], mscorlib",
        "$values": [
          {
            "$type": "NonVisuals.StreamDeck.StreamDeckButton, NonVisuals",
            "StreamDeckButtonName": 22,
            "Face": {
              "$type": "NonVisuals.StreamDeck.FaceTypeImage, NonVisuals",
              "FaceType": 2,
              "ConfigurationOK": true,
              "TextFont": "Consolas, 12pt",
              "FontColor": "Black",
              "BackgroundColor": "Black",
              "ImageFile": "C:\\Users\\cahil\\OneDrive\\Documents\\Git\\dcs\\f16\\ufc\\0.png",
              "StreamDeckButtonName": 22,
              "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}",
              "OffsetX": 0,
              "OffsetY": 0
            },
            "ActionForPress": null,
            "ActionForRelease": null,
            "HasConfig": true,
            "ActionType": 0,
            "FaceType": 2
          }
        ]
      },
      "IsVisible": true,
      "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}"
    }
  ]
}
```

**Comment by matt-cahill** [comment]
ah I see. If it is a sub path of the application it is written out as relative.
If this is to help with sharing profiles it need to apply when relative to the saved profile `.bindings` file.
```json
{
  "$type": "System.Collections.Generic.List`1[[NonVisuals.StreamDeck.StreamDeckLayer, NonVisuals]], mscorlib",
  "$values": [
    {
      "$type": "NonVisuals.StreamDeck.StreamDeckLayer, NonVisuals",
      "Name": "Home",
      "StreamDeckButtons": {
        "$type": "System.Collections.Generic.List`1[[NonVisuals.StreamDeck.StreamDeckButton, NonVisuals]], mscorlib",
        "$values": [
          {
            "$type": "NonVisuals.StreamDeck.StreamDeckButton, NonVisuals",
            "StreamDeckButtonName": 22,
            "Face": {
              "$type": "NonVisuals.StreamDeck.FaceTypeImage, NonVisuals",
              "FaceType": 2,
              "ConfigurationOK": true,
              "TextFont": "Consolas, 12pt",
              "FontColor": "Black",
              "BackgroundColor": "Black",
              "ImageFile": "C:\\Users\\cahil\\OneDrive\\Documents\\Git\\dcs\\f16\\ufc\\0.png",
              "StreamDeckButtonName": 22,
              "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}",
              "OffsetX": 0,
              "OffsetY": 0
            },
            "ActionForPress": null,
            "ActionForRelease": null,
            "HasConfig": true,
            "ActionType": 0,
            "FaceType": 2
          },
          {
            "$type": "NonVisuals.StreamDeck.StreamDeckButton, NonVisuals",
            "StreamDeckButtonName": 21,
            "Face": {
              "$type": "NonVisuals.StreamDeck.FaceTypeImage, NonVisuals",
              "FaceType": 2,
              "ConfigurationOK": true,
              "TextFont": "Consolas, 12pt",
              "FontColor": "Black",
              "BackgroundColor": "Black",
              "ImageFile": "C:\\Users\\cahil\\Saved Games\\DCS.openbeta\\DCSFlightPanels\\testbouton.png",
              "StreamDeckButtonName": 21,
              "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}",
              "OffsetX": 0,
              "OffsetY": 0
            },
            "ActionForPress": null,
            "ActionForRelease": null,
            "HasConfig": true,
            "ActionType": 0,
            "FaceType": 2
          },
          {
            "$type": "NonVisuals.StreamDeck.StreamDeckButton, NonVisuals",
            "StreamDeckButtonName": 23,
            "Face": {
              "$type": "NonVisuals.StreamDeck.FaceTypeImage, NonVisuals",
              "FaceType": 2,
              "ConfigurationOK": true,
              "TextFont": "Consolas, 12pt",
              "FontColor": "Black",
              "BackgroundColor": "Black",
              "ImageFile": ".\\testbouton.png",
              "StreamDeckButtonName": 23,
              "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}",
              "OffsetX": 0,
              "OffsetY": 0
            },
            "ActionForPress": null,
            "ActionForRelease": null,
            "HasConfig": true,
            "ActionType": 0,
            "FaceType": 2
          }
        ]
      },
      "IsVisible": true,
      "StreamDeckInstanceId": "\\\\?\\hid#vid_0fd9&pid_006c#6&27db50b8&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}"
    }
  ]
}
```

**Comment by matt-cahill** [comment]
the `file not found` part works perfectly though

**Comment by jdahlblom** [comment]
Do you execute DCSFP from other drive than where the image files where stored? If so the absolute path will be given.

**Comment by matt-cahill** [comment]
I suppose then when sharing you could tell people to only unzip in a folder next to your DCSFP install folder and then `..\\my_cool_profile\\image.png` would work.
I think it's going to be super complicated to have it relative to the save file isn't it?

**Comment by matt-cahill** [comment]
Slight regression in 4.0.52 where the "Home" and "Back" navigation images are saved in the profile as fully qualified paths again. 

e.g.
```
"ImageFile": "D:\\DCSTools\\DCSFlightpanels_x64_4.0.52_alpha\\\\StreamDeckGallery\\Symbols\\home_white.png",
```

**Comment by matt-cahill** [comment]
ok this is resolved again in 4.0.59
